internal/provider/metric_export_prometheus_config_resource.go

```diff
@@ -0,0 +1,377 @@
+	rawResponse, _ := json.MarshalIndent(apiObj, "", "  ")
+	tflog.Info(ctx, fmt.Sprintf("enable prometheus metric export response: %s", string(rawResponse)))
+
+	var state ClusterPrometheusMetricExportConfig
+	var targets types.Map
+	targets, diags = types.MapValueFrom(ctx, types.StringType, apiObj.GetTargets())
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		resp.Diagnostics.AddError(
+			"Error in parsing enable prometheus metric export",
+			fmt.Sprintf(
+				"There is an error in parsing enable prometheus metric export. response: %v",
+				apiObj,
+			))
+		return
+	}
+	loadPrometheusMetricExportIntoTerraformState(types.StringValue(apiObj.GetClusterId()), types.StringValue(string(apiObj.GetStatus())), targets, &state)
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+}
+
+func retryEnablePrometheusMetricExport(
+	ctx context.Context,
+	cl client.Service,
+	clusterID string,
+	cluster *client.Cluster,
+	apiObj *client.PrometheusMetricExportInfo,
+) retry.RetryFunc {
+	return func() *retry.RetryError {
+		body := &map[string]interface{}{}
+		apiResp, httpResp, err := cl.EnablePrometheusMetricExport(ctx, clusterID, body)
+		if err != nil {
+			apiErrMsg := formatAPIErrorMessage(err)
+			if (httpResp != nil && httpResp.StatusCode == http.StatusServiceUnavailable) ||
+				strings.Contains(apiErrMsg, "lock") {
+				// Wait for cluster to be ready.
+				clusterErr := retry.RetryContext(ctx, clusterUpdateTimeout,
+					waitForClusterReadyFunc(ctx, clusterID, cl, cluster))
+				if clusterErr != nil {
+					return retry.NonRetryableError(
+						fmt.Errorf("error checking cluster availability: %s", clusterErr.Error()))
+				}
+				return retry.RetryableError(fmt.Errorf("cluster was not ready - trying again"))
+			}
+
+			return retry.NonRetryableError(
+				fmt.Errorf("could not enable Prometheus metric export: %v", apiErrMsg),
+			)
+		}
+		*apiObj = *apiResp
+		return nil
+	}
+}
+
+func loadPrometheusMetricExportIntoTerraformState(
+	ID types.String, status types.String, targets types.Map, state *ClusterPrometheusMetricExportConfig,
+) {
+	state.ID = ID
+	state.Status = status
+	state.Targets = targets
+}
+
+func waitForPrometheusMetricExportStableFunc(
+	ctx context.Context,
+	clusterID string,
+	cl client.Service,
+	prometheusMetricExportInfo *client.PrometheusMetricExportInfo,
+) retry.RetryFunc {
+	return func() *retry.RetryError {
+		apiObj, httpResp, err := cl.GetPrometheusMetricExportInfo(ctx, clusterID)
+		if err != nil {
+			if httpResp != nil && httpResp.StatusCode < http.StatusServiceUnavailable {
+				return retry.NonRetryableError(fmt.Errorf(
+					"error getting Prometheus metric export info: %s", formatAPIErrorMessage(err)))
+			} else {
+				return retry.RetryableError(fmt.Errorf(
+					"encountered a server error while reading Prometheus metric export status - trying again"))
+			}
+		}
```

Comment:
Made a small change - `< http.StatusInternalServerError`

---

> How do we know this error is retryable

I think the assumption here is that all internal servers are temporary and retryable. Because this is a GET request with no URL params.

---

> I'm also curious what happens if the retry func times out after falling into this block over and over?

I think it reaches here and shows up as a request timeout: https://github.com/cockroachdb/terraform-provider-cockroach/blob/ef9d625265a359472c218ebe977ad943d8358927/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry/state.go#L260-L262

---

> In this block we've never actually logged the or returned the actual error.

It would be helpful to log this error condition. I've added a log line that denotes that there is an internal server error so the request will be retried.